domain/secret/service/service.go

```diff
@@ -315,14 +319,15 @@ func (s *SecretService) CreateCharmSecret(ctx context.Context, uri *secrets.URI,
 			}
 			return errors.Trace(err)
 		}
-		err = s.secretState.CreateCharmApplicationSecret(ctx, params.Version, uri, params.CharmOwner.ID, p)
-	} else {
-		err = s.secretState.CreateCharmUnitSecret(ctx, params.Version, uri, params.CharmOwner.ID, p)
 	}
-	if errors.Is(err, secreterrors.SecretLabelAlreadyExists) {
-		return errors.Errorf("secret with label %q is already being used", *params.Label)
-	}
-	if err != nil {
+
+	if err = s.secretState.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		owner := secrets.Owner{
+			ID:   params.CharmOwner.ID,
+			Kind: secrets.OwnerKind(params.CharmOwner.Kind),
+		}
+		return s.createSecret(ctx, params.Version, uri, owner, p)
+	}); err != nil {
```

Comment:
This really should be broken out into two statements, not trying to shove it into an if assignment.